### PR TITLE
Fixes 3832: update date type in /snapshots/for_date

### PR DESCRIPTION
**Pull request by xbhouse**
## Summary

- Updates the date in the request to /snapshots/for_date to be an alias of type time.Time
- Date can be formatted like either YYYY-MM-DD or 2024-05-02T00:00:00-04:00
- Frontend change for this as well to ensure template creation doesn't fail on the Set up date step, PR [here](https://github.com/content-services/content-sources-frontend/pull/256). <-- this is not needed anymore for this PR since both formats work, but will be once we remove support for YYYY-MM-DD in a following PR

## Testing steps

- Test the /snapshots/for_date endpoint with a request like this, this should work:
```
{
  "date": "2024-04-28T00:00:00-04:00",
  "repository_uuids": ["1510694a-1bfa-479a-b596-4377261f3135"]
}
```
- This should also work:
```
{
  "date": "2024-04-28",
  "repository_uuids": ["1510694a-1bfa-479a-b596-4377261f3135"]
}
```

## Checklist

- [ ] Tested with snapshotting feature disabled and pulp server URL not configured if appropriate

### CODE REVIEW

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/HMS-3832

**Comment by mayurilahane** [comment]
`In [9]: app.content_sources.rest_client.snapshots_api.list_snapshots_by_date({
   ...:   "date": "2024-04-28T00:00:00-04:00",
   ...:   "repository_uuids": ["a90f22f8-ccc4-4ef7-9aa6-80430576d73d"]
   ...: })
2024-05-06 11:02:35.430 [    INFO] [iqe.base.rest_client] REST: METHOD=POST, request_id=22de383d-8c5e-42fb-8c36-00c6fd5a955b, params=[]
Out[9]: 
{'data': [{'is_after': True,
           'match': {'added_counts': {'rpm.package': 33},
                     'content_counts': {'rpm.package': 33},
                     'created_at': '2024-05-06T15:01:15.295761Z',
                     'removed_counts': {},
                     'repository_path': '28fbe535/a90f22f8-ccc4-4ef7-9aa6-80430576d73d/99210df9-49e3-4c4f-9da2-5d3d22920da2',
                     'url': '',
                     'uuid': '5f01b5d5-4a29-4de8-8242-2b6617f7f877'},
           'repository_uuid': 'a90f22f8-ccc4-4ef7-9aa6-80430576d73d'}]}
`

**Comment by mayurilahane** [comment]
And got 400 error 
for the old date type 
image

**Comment by jlsherrill** [comment]
⚠️ This task currently requires qe-approval, but this PR does not fully resolve the task.  Please contact QE to determine appropriate testing required.

**Comment by Andrewgdewar** [comment]
> And got 400 error for the old date type image

This was updated to now allow both date types (temporarily).

**Review by Andrewgdewar — APPROVED** [review]
Tested with both date formats, working well.
Ensure that @croissanne is told that we switched to this date-type. 
She originally wanted to use RFC 3339 as the format format for her blueprints api snapshot date, we should likely get her to switch her api over to that date as well. 
Making updates to the image-builder front-end snapshots integration to support that.

File: pkg/api/snapshots.go [hunk 1]
```diff
@@ -16,7 +17,34 @@ type SnapshotResponse struct {
...
 type ListSnapshotByDateRequest struct {
 	RepositoryUUIDS []string `json:"repository_uuids"` // Repository UUIDs to find snapshots for
-	Date            string   `json:"date"`             // Exact date to search by.
+	Date            Date     `json:"date"`             // Exact date to search by.
+}
+
+type Date time.Time
+
+func (d Date) MarshalJSON() ([]byte, error) {
+	return json.Marshal(time.Time(d).Format(time.RFC3339))
```
Comment (rverdile) [inline pkg/api/snapshots.go:26]: I don't think the marshal function should force a certain format. I think it should just use whatever format was being used for date.

Comment (rverdile) [reply pkg/api/snapshots.go:26]: Actually that doesn't make sense. Ignore me.

**Comment by xbhouse** [comment]
> > And got 400 error for the old date type image
> 
> This was updated to now allow both date types (temporarily).

@mayurilahane ^

**Comment by mayurilahane** [comment]
image

**Comment by mayurilahane** [comment]
image

**Comment by mayurilahane** [comment]
LGTM!

**Pull request merged**